core/src/main/java/com/graphhopper/routing/template/RoundTripRoutingTemplate.java

```diff
@@ -0,0 +1,180 @@
+/*
+ *  Licensed to GraphHopper and Peter Karich under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for 
+ *  additional information regarding copyright ownership.
+ * 
+ *  GraphHopper licenses this file to you under the Apache License, 
+ *  Version 2.0 (the "License"); you may not use this file except in 
+ *  compliance with the License. You may obtain a copy of the License at
+ * 
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.template;
+
+import com.graphhopper.GHRequest;
+import com.graphhopper.GHResponse;
+import com.graphhopper.PathWrapper;
+import com.graphhopper.routing.*;
+import com.graphhopper.routing.util.AvoidEdgesWeighting;
+import com.graphhopper.routing.util.DefaultEdgeFilter;
+import com.graphhopper.routing.util.EdgeFilter;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.routing.util.tour.SinglePointTour;
+import com.graphhopper.routing.util.tour.TourStrategy;
+import com.graphhopper.storage.index.LocationIndex;
+import com.graphhopper.storage.index.QueryResult;
+import com.graphhopper.util.*;
+import com.graphhopper.util.shapes.GHPoint;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of calculating a route with one or more round trip (route with identical start and
+ * end).
+ *
+ * @author Peter Karich
+ */
+public class RoundTripRoutingTemplate implements RoutingTemplate
+{
+    private final int maxRetries;
+    private final GHRequest ghRequest;
+    private final GHResponse ghResponse;
+    private PathWrapper altResponse;
+    private final LocationIndex locationIndex;
+    // result from lookup
+    private List<QueryResult> queryResults;
+    // result from route
+    private List<Path> pathList;
+
+    public RoundTripRoutingTemplate( GHRequest request, GHResponse ghRsp, LocationIndex locationIndex, int maxRetries )
+    {
+        this.ghRequest = request;
+        this.ghResponse = ghRsp;
+        this.locationIndex = locationIndex;
+        this.maxRetries = maxRetries;
+    }
+
+    @Override
+    public List<QueryResult> lookup( List<GHPoint> points, FlagEncoder encoder )
+    {
+        double distanceInMeter = ghRequest.getHints().getDouble("round_trip.distance", 1000);
+        long seed = ghRequest.getHints().getLong("round_trip.seed", 0L);
+        GHPoint start = ghRequest.getPoints().get(0);
```

Comment:
As far as I can see, we don't check if there is at least one point in ponts. We should add this somewhere.